source/D2Game/include/GAME/Clients.h

```diff
@@ -166,11 +166,11 @@ struct D2ClientStrc
 	uint8_t nClassId;							//0x008
 	uint8_t unk0x09;							//0x009
 	uint16_t nSaveFlags;						//0x00A D2ClientSaveFlags
-	uint8_t unk0x0C;							//0x00C
+	uint8_t nCharTemplate;						//0x00C
 	char szName[16];							//0x00D
 	char szAccount[16];							//0x01D
 	uint8_t unk0x2D[51];						//0x02D
-	int32_t unk0x60;							//0x060
+	int32_t nDatabaseCharacterId;				//0x060
```

Comment:
IIRC some of the Blizzard asserts call this the PlayerToken (though it is a "database id"), it is unused by D2Game however (though its passed/set to/in D2Server).

Reply:
player token is a bit opaque, as it's mostly used to validate that the save that has been provided matches the one it expects (which is set at VerifyJoin time).
I guess I want to make it clear that this is used for save loading.

Reply:
The issue here is that if we know the Blizzard original naming, then we use the Blizzard naming, anything about usage should then be a comment etc on the field.

Reply:
I don't think it is.
The pointer to this value is passed to the pfFindPlayerToken callback. The token itself is 2nd argument to that call. This id is the 5th argument, which gets set when the realm server/game server "finds the token".
Then when the game server asks to provide a save for that client, it expects to get the same id in the call providing the save.

Reply:
So nCharacterId? Or what should I call it?

Reply:
I'd rather have it be named `nUnknownBnetId` until I figure out what it is exactly if that's OK with you?

Reply:
I'm fairly certain it's related to the character, as it's used when character save it sent, so nUnknownCharacterId  ?

Reply:
So my investigation led to the following:

- Client => MCP: User requests to join game
- MCP => GS: Marshalls the request. Game server temporarily stores packet as a `PlayerToken` with: 
  - `MCP Id`
  - `TokenId`
  - `Character Name`
  - `Account Name`
  - `GameId`
  - `Our unknown ID`
- GS => MCP: Acknowledges it received the request
- GS D2Game receives `D2CLTSYS_JOINGAME` packet for game join
  - Calls `GAME_VerifyJoinGame` and thus `ptEventCallbackTable->fpFindPlayerToken` which returns: `szAccountName` and `Our unknown ID`, and removes the token. (It's more like a `FindAndRemove` than find)
  - Calls `CLIENTS_AddToGame` which creates the client with `Our unknown ID`
- CharacterDBServer =>GS: 
  - Packet has:
    - Client ID
    - Save data
    - Save chunk Size
    - `Our unknown ID`
  - GS handles it by calling GAME_ReceiveDatabaseCharacter and checks that `Our unknown ID` matches the one it has for the client. It does NOT check the value from the save itself
- When GS saves a character, it saves `Our unknown ID`
  - Then calls `gpD2EventCallbackTable_6FD45830->pfSaveDatabaseCharacter` which sends `Our unknown ID` to the DBServer
- GS D2Game calls `gpD2EventCallbackTable_6FD45830->pfLeaveGame` with `Our unknown ID` (but only in recent versions of the game, not in 1.00)

So what we can see:
- The ID is never stored
- The ID is checked against an ID received first received by the MCP server, and then used during transactions with the CharDB. This could be changing every game.

- What it could be:
  - A character db server id ?
  - A server wide game transaction id ?
  - Something related to the account itself (since FindPlayerToken returns both the account name and `Our unknown ID`)
  - some kind of token for the save to avoid dupes/handle rollbacks ?

Sadly, it is never seen on the client side, this ID only exists between GS, MCP and the DB server...
So it's a bit hard to guess.

What do you think about something along the lines of `CharSaveTransactionToken` ?